website/docs/language/expressions/custom-conditions.mdx

```diff
@@ -0,0 +1,353 @@
+
+Use `precondition` and `postcondition` blocks to create custom rules for resources, data sources, and outputs.
+
+Terraform checks a precondition _before_ evaluating the object it is associated with and checks a postcondition _after_ evaluating the object. Terraform evaluates custom conditions as early as possible, but must defer conditions that depend on unknown values until the apply phase. Refer to [Delaying Evaluation Until Apply](#delaying-evaluation-until-apply) for more details.
+
+### Usage
+
+Each precondition and postcondition requires a [`condition` argument](#condition-expressions). This is an expression that must return `true` if the conditition is fufilled or `false` if it is invalid. The expression can refer to any other objects in the same module, as long as the references do not create cyclic dependencies. Resource postconditions can also use the symbol `self` to refer to attributes of each instance of the resource where they are configured. For example, `self.private_dns` refers to the `private_dns` attribute of each instance of the containing resource.
+
+If the condition evaluates to `false`, Terraform will produce an [error message](#error-messages) that includes the result of the `error_message` expression. If you declare multiple preconditions or postconditions, Terraform returns error messages for all failed conditions.
+
+The following example uses a postcondition to detect if the caller accidentally provided an AMI intended for the wrong system component. This might otherwise be detected only after booting the EC2 instance and noticing that the expected network service is not running.
+
+``` hcl
+data "aws_ami" "example" {
+  id = var.aws_ami_id
+
+  lifecycle {
+    # The AMI ID must refer to an existing AMI that has the tag "nomad-server".
+    postcondition {
+      condition     = self.tags["Component"] == "nomad-server"
+      error_message = "The selected AMI must be tagged with the
+      Component value \"nomad-server\"."
+    }
+  }
+}
+```
+
+You can add `precondition` and `postcondition` blocks in the following locations within your configuration.
+
+#### Resources and Data Sources
+
+The `lifecycle` block inside a `resource` or `data` block can include both `precondition` and `postcondition` blocks. Terraform evaluates these blocks as follows:
+
+- Preconditions after evaluating existing `count` and `for_each` arguments. This lets Terraform evaluate the precondition separately for each instance and then make `each.key`, `count.index`, etc. available to those conditions.
+- Preconditions before evaluating the resource's configuration arguments. Preconditions can take precedence over argument evaluation errors.
+- Postconditions after planning and applying changes to a managed resource, or after reading from a data source. Postcondition failures prevent changes to other resources that depend on the failing resource.
```

Comment:
The first time I read this I kinda ran aground at these opening sentences like "Preconditions before evaluating the resource's configuration arguments", because alone they are ungrammatical... my brain said "this sentence no verb" and I had to concentrate hard to realize what was meant.

Now I've read it a few times I realize this is in the context of "Terraform evaluates these blocks as follows:" and so the verb is implied to be "evaluates", but for me at least the parsing distance between that opening sentence and these bullet points was too far for me to retain that context to understand these sentences.

I'm not sure what to suggest that doesn't make it _super wordy_ and avoids using the passive voice. My best effort is just to add "Terraform evaluates" to the beginning of each, which is redundant with the introductory sentence but seems to be the smallest addition I could make to these sentences to make them read as full sentences.

This might be a me-specific quirk, so maybe not a big deal. :man_shrugging:

Reply:
heh I had literally the EXACT same thoughts and so I wanted to test drive this. But I'm glad to know you tripped on it because I was worried about that....and now I'm going to change it back to "Terraform evaluates".  Redundant, but I might bold the relevant word in each or something similar....

Reply:
okay I'm trying a thing here where I put back "Terraform evaluates" but to try to make the condition type more salient when skimming, I used the block name and put it in code format. What do you think about this - better?
<img width="954" alt="Screen Shot 2022-04-07 at 12 00 14 PM" src="https://user-images.githubusercontent.com/83350965/162246866-ea6ffe09-c2b2-4edd-85f2-a608c2ccf3d8.png">